README.md
# EGD Simulator

## Introduction

Ethernet Global Data (EGD) is a mechanism that enables one CPU, referred to as a producer, to
share a portion of its internal reference memory with one or more other CPUs, referred to as
consumers, at a regularly scheduled periodic rate. Such a snapshot of internal reference
memory, mediated by an Ethernet interface, is referred to as an exchange. An exchange is
identified by a unique combination of three identifiers:
• The Producer ID (the producer's IP address)
• The Exchange ID (the exchange's identifier)
• The Adapter Name (the Ethernet interface identifier) 

The EGD protocol is used in many contexts of Industrial monitoring and control and is backed by GE. 

Please refer to [this document](http://geplc.com/downloads/Labs/GFS-384%20M09%20EGD.pdf) for further information.

## EGD Simulator

Our EGD Simulator is an effort to create a software that simulates typical EGD producers that can be found as control hardware like for example [Mark VI e (C)](http://www.geautomation.com/products/mark-vie) created by GE Automation (R).

EGD Simulator is a Python 2.7 and 3.x program with a web frontend.

Once the software is installed it can be executed with the command from a command line prompt:

```
python egdsimulator_http.py
```
At that point, opening a browser and pointing to the address *http://localhost:8000* is sufficient to get to the configuration interface.

It is also possible to change the port of the configuration interface. In order to do so the simulator has to be started with the following command line:

```
python egdsimulator_http.py -p NEWPORT
```
From the configuration interface it's possible to upload a CSV representing the setup of the simulator. Currently the simulator supports the CSV format obtained from exporting a setup from the MARK VI e control software. In addition, a simplified CSV format that contains only the four relevant information fields for every tag can be uploaded.

Two example files for both formats are supplied in the _resource_ folder.

Separator and initial number of lines to be skipped for the CSV parser may be specified before uploading.

Example MARK VI export format is:

```
#(V05.02.04C) EGD Point List Report for example,,,,,,,,,,,,,,,,
Entry No,Variable,Point Name,Point Offset,Direction,Data Type,CIMPLICITY Resource,Units,Producer ID,Period(ms),Exchange Name,Destination,Description,Alias,Second Language Description,Path,Format Specification
617,FAKEBOOL1,EGD.ProducedPages.FastEGD.Exchange3.Point_23,03.0003.1,ReadOnly,BOOL,,nd,1701122220,10,Exchange3,172.16.111.255,fake bool 1,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:FastEGD|EgdExchange:Exchange3|EgdVariable:a,N_D
618,FAKEBOOL2,EGD.ProducedPages.FastEGD.Exchange3.Point_23,03.0003.2,ReadOnly,BOOL,,nd,1701122220,10,Exchange3,172.16.111.255,fake bool 2,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:FastEGD|EgdExchange:Exchange3|EgdVariable:b,N_D
619,FAKEREAL1,EGD.ProducedPages.FastEGD.Exchange3.Point_30,03.0032.0,ReadOnly,REAL,,lb/hr,1701122220,10,Exchange3,172.16.111.255,fake real 1,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:FastEGD|EgdExchange:Exchange3|EgdVariable:c,PPH
620,FAKEREAL2,EGD.ProducedPages.FastEGD.Exchange3.Point_31,03.0036.0,ReadOnly,REAL,,,1701122220,10,Exchange3,172.16.111.255,fake real 2,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:FastEGD|EgdExchange:Exchange3|EgdVariable:d,
627,FAKEDINT,EGD.ProducedPages.FastEGD.Exchange3.Point_40,03.0072.0,ReadOnly,DINT,,nd,1701122220,10,Exchange3,172.16.111.255,fake dint,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:FastEGD|EgdExchange:Exchange3|EgdVariable:e,N_D
649,FAKEINT,EGD.ProducedPages.FastEGD.Exchange3.Point_23,03.0004.0,ReadOnly,INT,,nd,1701122220,10,Exchange3,172.16.111.255,fake int,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:FastEGD|EgdExchange:Exchange3|EgdVariable:f,N_D
1334,FAKEUDINT,EGD.ProducedPages._Status.Exchange1.Point_464,01.0292.0,ReadOnly,UDINT,,,1701122220,250,Exchange1,172.16.111.255,fake udint,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:_Status|EgdExchange:Exchange1|EgdVariable:g,
1498,FAKEUINT,EGD.ProducedPages._Status.Exchange1.Point_400,01.0060.0,ReadOnly,UINT,,,1701122220,250,Exchange1,172.16.111.255,fake uint,,,System:PTT_1609256_59|Device:T1|EGD:|EgdPageList:ProducedPages|EgdPage:_Status|EgdExchange:Exchange1|EgdVariable:h,

```

Example simplified CSV format is:

```
Producer,Exchange,Address,Type
1,01,0000.0,UINT
1,01,0002.0,UINT
1,01,0004.0,UINT
2,03,0000.1,BOOL
2,03,0000.2,BOOL
2,03,0010.0,UINT
```
Once the setup CSV has been uploaded the simulator can be started through the play button and paused through the pause button.

It is possible to upload a different CSV while the simulator is running. This just changes the layout of the tags but keeps the simulator going.

When the simulator is in paused mode, it's possible to change the value generation policy to _fixed_. This forces the simulator to generate always the same value for each tag. If _fixed_ is not set, random values will be generated according to the tag type.

It is also possible, in paused mode, to change the UDP address to which the simulator publishes the Exchange values.

In order to overview the EGD structure and the values produced by the simulator, the web page shows a table with all Producers, Exchanges, Tags and generated values.

In order to see the changes in the readings it's just sufficient to reload the webpage. 

## Version

Version 0.5


CHANGES
EGD Simulator Changelog
===================
0.5 
---------------
Better packaging


0.1 (unreleased)
----------------

- First release.


egdsimulator.py
#!/bin/python

import threading
import time
import datetime
import socket
import random
import struct
import egdmodel

class EGDSimulator():

		def __init__(self, egd, address="127.0.0.255", fixed = False):
			self.egd = egd
			self.sock = None
			self.socklock = threading.Lock()
			self.count = 0
			self.address = address
			self.fixed = fixed

		def computeValue(self, tag):
			if tag.type in ["REAL","LREAL"]:
				tag.lastvalue = tag.lastvalue if tag.lastvalue != None else (-1000.0 + random.random() * 2000.0)
				tag.lastvalue += 0 if self.fixed else -10 + random.random() * 20
			elif tag.type == "DINT":
				tag.lastvalue = tag.lastvalue if tag.lastvalue != None else random.randrange(-1000,1000)
				tag.lastvalue += 0 if self.fixed else random.randrange(-10,11) 
			elif tag.type == "INT":
				tag.lastvalue = tag.lastvalue if tag.lastvalue != None else random.randrange(-100,100)
				tag.lastvalue += 0 if self.fixed else random.randrange(-1,2)
			elif tag.type == "UDINT":
				tag.lastvalue = tag.lastvalue if tag.lastvalue != None else random.randrange(0,2000)
				tag.lastvalue += 0 if self.fixed else random.randrange(0,21)
			elif tag.type == "UINT":
				tag.lastvalue = tag.lastvalue if tag.lastvalue != None else random.randrange(0,200)
				tag.lastvalue += 0 if self.fixed else random.randrange(0,3)
			elif tag.type == "BOOL":
				tag.lastvalue = tag.lastvalue if tag.lastvalue != None else random.randrange(0,2)
				tag.lastvalue = tag.lastvalue if (self.fixed or random.randrange(0,10000) < 9998) else not tag.lastvalue
			
			return tag.lastvalue

		def computeValues(self, exchange):
			values = []
			lasttag = None
			for tag in exchange.tags:
				v = self.computeValue(tag)
				v = v if tag.type != "BOOL" else v << int(tag.offsetbit)
				if lasttag != None and lasttag.offsetbyte == tag.offsetbyte:
					values[len(values)-1] = values[len(values)-1] | v
				else:
					values.append(v)
					lasttag = tag
					
			return values

		def produce(self, producer, exchange):
			if self.isRunning():
				self.count += 1
				#record last time stamp for dumping purposes by using lasttimestamp field of egd exchange
				tm = time.localtime()
				exchange.lasttimestamp = "" + str(self.count) + " at " + str(tm.tm_hour).zfill(2) + ":" + str(tm.tm_min).zfill(2) + ":" + str(tm.tm_sec).zfill(2)
				hdr = egdmodel.EGDHeader.newHeaderFor(producer.producerid, exchange.exchangenumber, self.count).header
				payload = struct.pack(exchange.codec, *self.computeValues(exchange))
				self.sock.sendto(hdr+payload, (self.address, egdmodel.UDP_DATA_PORT))
				threading.Timer(float(exchange.period) / 1000.0, self.produce, [producer, exchange]).start()

		def isRunning(self):
			with self.socklock:
				return self.sock != None

		def stop(self):
			with self.socklock:
				if self.sock != None: 
					self.sock.close()
					self.sock = None

		def start(self, producerid = None):
			with self.socklock:
				self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
			for producerid in self.egd.producers:
				producer = self.egd.producers[producerid]
				for exchangenumber in producer.exchanges:
					exchange = producer.exchanges[exchangenumber]
					if self.fixed: self.computeValues(exchange)
					threading.Timer(float(exchange.period) / 1000.0, self.produce, [producer, exchange]).start()



MKVIe/__init__.py
import csv

class Parser(object):

	def __init__(self, model):
		self.model = model
		pass

	def parse(self, csvdata, delimiter=",", skiplines = 2):
		self.delimiter = delimiter
		self.skiplines = skiplines

		if csvdata == None or csvdata == "":
			return
		
		self.reader = None
		if not (hasattr(csvdata, 'read') and hasattr(csvdata, 'write')):
			self.reader = csv.reader(csvdata.split("\n"), delimiter=delimiter)
		else:
			self.reader = csv.reader(csvdata, delimiter=delimiter)
		
		for i in range(self.skiplines):
			next(self.reader)

		for row in self.reader:
			if row != None and len(row) > 11:
				producerid = row[8]
				address = row[3]
				tagtype = row[5]
				period = row[9]
				destination = row[12]

				producer = self.model.getProducer(producerid)
				if producer == None:
					producer = self.model.addProducer(producerid, destination)

				exchangenumber, byteoffset, bitoffset = address.split(".")
				exchange = producer.getExchange(exchangenumber)
				if exchange == None:
					exchange = producer.addExchange(exchangenumber, period)
				
				#print("Adding tag to exchange",exchange.exchangenumber,tagtype,byteoffset, bitoffset)
				t = exchange.addTagFromAddressParts(tagtype, byteoffset, bitoffset)
				#print(t.dump())
		self.model.buildCodecs()


egdsimulator_http.py
import cgi, sys, getopt
try:
	from BaseHTTPServer import BaseHTTPRequestHandler
	import SocketServer
except:
	from http.server import BaseHTTPRequestHandler
	import socketserver as SocketServer

import egdmodel
import MKVIe, Custom
import egdsimulator

class MyHandler(BaseHTTPRequestHandler):

	def buildTag(self, tag):
		return"""<tr class="tag"><td>{}</td><td>{}</td><td>{}</td></tr>""".format(
				tag.offsetbyte+"."+tag.offsetbit, 
				tag.type, 
				tag.lastvalue if tag.lastvalue != None else "")

	def buildExchange(self, exchange):
		print(exchange.dump())
		return"""
			<tr class="exchange">
				<td>Exchange {}</td>
				<td>Period {} (ms)</td>
				<td>Msg {}</td>
				<td>{} tags</td>
				<td class="toggle"><label></label></td>
			</tr>
			{}
		""".format(
					exchange.exchangenumber, exchange.period, 
					exchange.lasttimestamp if exchange.lasttimestamp != None else "", 
					len(exchange.tags),
					"".join(list(map(lambda t: self.buildTag(t), exchange.sortedTags()))))

	def buildProducer(self, producer):
		return"""
			<tr class="producer"><td>Producer {}</td></tr>
			{}
		""".format(
					producer.producerid, 
					"".join(map(lambda e: self.buildExchange(producer.exchanges[e]), producer.exchanges.keys())))

	def buildTable(self, configuration):
		return"""
			<table><tbody>{}</tbody></table>
		""".format(
					"".join(list(
							map(lambda p: self.buildProducer(configuration.producers[p]), configuration.producers.keys()))
						)
			)

	def buildPage(self, simulator, state):
		page = (open("resources/index.html")).read()
		patches = { 
			"error" : state["error"],
			"errorvisible" : "" if state["error"] != None else "invisible",
  		"delimiter" : state["delimiter"],
  		"skiplines" : state["skiplines"],
  		"address" : simulator.address,
  		"fixed" : "checked" if simulator.fixed else "",
			"displaystart" : "inactive" if simulator.isRunning() else "active",
			"displaystop" : "active" if simulator.isRunning() else "inactive",
			"egd" : self.buildTable(simulator.egd) if simulator.egd != None else ""
		}
		page = page.format(**patches)
		state["error"] = None
		return page
	
	def buildDocumentationPage(self):
		page = (open("resources/documentation.html")).read()
		patches = { 
			"license" : open("LICENSE").read().replace("\n", "<br/>"),
			"authors" : open("AUTHORS").read().replace("\n", "<br/>"),
			"manual" : open("README.md").read().replace("\n", "<br/>")
		}
		page = page.format(**patches)
		return page
	
	def returnPage(self, page):
		self.send_response(200)
		self.send_header('Content-type', 'text/html')
		self.end_headers()
		if sys.version_info < (3,):
			self.wfile.write(page)
		else:
			self.wfile.write(bytes(page, "UTF-8"))

	def returnResource(self, resource, mediatype):
		self.send_response(200)
		self.send_header('Content-type', (mediatype if mediatype != None else ""))
		self.end_headers()
		self.wfile.write(open(resource, "rb").read())

	def redirect(self, location):
		self.send_response(301)
		self.send_header('Location',location)
		self.end_headers()

	def do_GET(self):

		global simulator
		global state

		if self.path == "/":
			self.returnPage(self.buildPage(simulator, state))
			return
		if self.path == "/logo":
			self.returnResource("resources/logonubisware.png", "image/png")
			return
		if self.path == "/favicon.ico":
			self.returnResource("resources/favicon.ico", "image/ico")
			return
		if self.path == "/documentation":
			self.returnPage(self.buildDocumentationPage())
			return
		else:
			self.redirect("/")
			return

	def parseMultipartPy2(self):
		ctype, pdict = cgi.parse_header(self.headers.getheader('content-type'))
		if not ctype == "multipart/form-data":
			self.redirect("/")
			return

		query = cgi.parse_multipart(self.rfile, pdict)

		mkvie = query.get("mkvie")[0] if query.get("mkvie")[0].decode("utf-8-sig") != "" else None
		custom = query.get("custom")[0] if query.get("custom")[0].decode("utf-8-sig") != "" else None

		state["delimiter"] = query.get("delimiter")[0]
		state["skiplines"] = int(query.get("skiplines")[0])
		
		return (mkvie, custom)
			
	def parseMultipartPy3(self):
		global state
		ctype, pdict = cgi.parse_header(self.headers.get('content-type'))
						
		if not ctype == "multipart/form-data":
			self.redirect("/")
			return

		pdict["boundary"] = bytes(pdict["boundary"], "utf-8")
			
		query = cgi.parse_multipart(self.rfile, pdict)
		
		mkvie = query.get("mkvie")[0] if query.get("mkvie")[0].decode("utf-8-sig") != "" else None
		custom = query.get("custom")[0] if query.get("custom")[0].decode("utf-8-sig") != "" else None

		state["delimiter"] = (query.get("delimiter")[0]).decode("utf-8")
		state["skiplines"] = int(query.get("skiplines")[0])
		
		return (mkvie, custom)

	def do_POST(self):

		global simulator
		global state

		if self.path == "/start":
			ctype, pdict = cgi.parse_header(self.headers.get('content-type'))
			
			try:
				pdict["boundary"] = bytes(pdict["boundary"], "utf-8") #fix 2 to 3 bug
			except:
				pdict["boundary"] = bytes(pdict["boundary"])
				
			if not ctype == "multipart/form-data":
				self.returnPage(self.buildPage())
				return
			query = cgi.parse_multipart(self.rfile, pdict)

			simulator.fixed = query.get("fixed") != None and query.get("fixed")[0] != ""

			if query.get("address")[0] != None and query.get("address")[0] != "":
				simulator.address = (query.get("address")[0]).decode("utf-8")

			if not simulator.isRunning():
				simulator.start()
			self.redirect("/")

		if self.path == "/stop":
			if simulator.isRunning():
				simulator.stop()
			self.redirect("/")

		if self.path == "/configure":

			mkvie = None
			custom = None
			if sys.version_info < (3,):
				mkvie, custom = self.parseMultipartPy2()
			else:
				mkvie, custom = self.parseMultipartPy3()
			
			wasrunning = simulator.isRunning()
			if wasrunning: simulator.stop()

			if mkvie != None and mkvie != "" and mkvie != b"":
				print("Parsing MKVIE")
				newegd = egdmodel.EGDConfiguration()
				try:
					MKVIe.Parser(newegd).parse(mkvie, state["delimiter"], state["skiplines"])
				except Exception as e:
					state["error"] = "Unable to parse MKVIe CSV: " + str(e) + str(sys.exc_info()[2].tb_lineno)
					self.redirect("/")

				simulator.egd = newegd
				print(simulator.egd.dump())

			elif custom != None and custom != "" and custom != b"":
				print("Parsing Custom")
				newegd = egdmodel.EGDConfiguration()
				try:
					Custom.Parser(newegd).parse(custom, state["delimiter"], state["skiplines"])
				except Exception as e:
					state["error"] = "Unable to parse Custom CSV: " + str(e) + str(sys.exc_info()[2].tb_lineno)
					self.redirect("/")

				simulator.egd = newegd
				print(simulator.egd.dump())

			if wasrunning: 
				simulator.start()

			self.redirect("/")

			return

		else:
			self.redirect("/")
			return

simulator = None
state = {
	"skiplines" : 2,
	"delimiter" : ",",
	"error" : None
}

def dumpUsage():
	print("python " + sys.argv[0] + ' [-p <port>] [-a <initial send address>]')

def boot(argv):
	global simulator
	global state
	HTTP_PORT = 8000
	INIT_ADDRESS = "127.0.0.255"
	#INIT_CSV = None
	try:
		opts, args = getopt.getopt(argv[1:],"hp:a:")
		for opt, arg in opts:
			if opt == "-h":
				dumpUsage()
				sys.exit(0)
			elif opt == "-p":
				HTTP_PORT = int(arg)
			elif opt == "-a":
				INIT_ADDRESS = int(arg)

	except getopt.GetoptError:
		dumpUsage()
		sys.exit(2)

	#Initialize with empty configuration and default values for fixed and address
	simulator = egdsimulator.EGDSimulator(egdmodel.EGDConfiguration(), INIT_ADDRESS, False)

	httpd = SocketServer.TCPServer(("", HTTP_PORT), MyHandler, bind_and_activate=False)

	httpd.allow_reuse_address = True
	httpd.daemon_threads = True
	try:
		print("\nCreating HTTP server...")
		httpd.server_bind()
		httpd.server_activate()
		print(sys.argv[0] + " HTTP server is listening on port " + str(HTTP_PORT) + "...")
		httpd.serve_forever()
	except KeyboardInterrupt as e:
		print("Interrupted by user...")
	except Exception as e:
		print(e)
		print(traceback.format_exc())

	try:
		print("Stopping simulator...")
		simulator.stop()
		print("Stopping HTTP server...")
		httpd.shutdown()
		httpd.server_close()
	except Exception as e:
		print(e)
		print(traceback.format_exc())

	sys.exit(1)

if __name__ == "__main__":
	boot(sys.argv)
